In [1]:
import pandas as pd
import numpy as np

In [2]:
train_data = pd.read_csv("../data/train.csv")
validation_data = pd.read_csv("../data/validation.csv")
test_data = pd.read_csv("../data/test.csv")

print(f"Shape of train_data is : {train_data.shape}")
print(f"Shape of validation_data is : {validation_data.shape}")
print(f"Shape of test_data is : {test_data.shape}")
train_data.head()

Shape of train_data is : (7827, 14)
Shape of validation_data is : (870, 14)
Shape of test_data is : (967, 14)


,Unnamed,team,opp,host,year,month,toss,day_match,bat_first,format,fow,score,rpo,result
0,4901,australia,south_africa,sri_lanka,2012,sep,1,0,0,1,5,146,7.30,1
1,1910,india,australia,india,2020,jan,0,0,1,0,6,340,6.80,1
2,5735,canada,scotland,scotland,2009,jul,1,1,0,0,4,286,5.72,1
3,2136,australia,england,australia,1987,jan,1,1,1,0,6,225,4.50,1
4,4371,new_zealand,pakistan,uae,2009,nov,0,0,0,1,5,153,7.65,0


In [3]:
import sys
import os
sys.path.insert(0, os.path.abspath('..'))  # Adjust path as needed
from utils import convertCategoricalColumnsToOneHotEncoding

train_data_encoded, val_data_encoded, test_data_encoded = convertCategoricalColumnsToOneHotEncoding(train_data, validation_data, test_data)

In [4]:
from utils import pre_process

# Step 2: Encode the columns["Unnamed", "team", "opp", "host", "month"]
continuous_cols = ["Unnamed", "year", "fow", "score", "rpo"]

trainDataX, trainDataY, validationDataX, validationDataY, testDataX, testDataY =  pre_process(train_data_encoded, val_data_encoded, test_data_encoded, categorical_cols = [], continuous_cols = continuous_cols)
trainDataX.head()

,Unnamed,bat_first,day_match,format,fow,host_australia,host_bangladesh,host_bermuda,host_canada,host_england,...,team_pakistan,team_scotland,team_south_africa,team_sri_lanka,team_uae,team_usa,team_west_indis,team_zimbambave,toss,year
0,0.023176,0,0,1,-1.226069,0,0,0,0,0,...,0,0,0,0,0,0,0,0,1,0.527774
1,-1.050706,1,0,0,-0.742084,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,1.215295
2,0.322613,0,1,0,-1.710053,0,0,0,0,0,...,0,0,0,0,0,0,0,0,1,0.269953
3,-0.969563,1,1,0,-0.742084,1,0,0,0,0,...,0,0,0,0,0,0,0,0,1,-1.620731
4,-0.167114,0,0,1,-1.226069,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0.269953


In [5]:
new_categorical_cols = list(trainDataX.columns[trainDataX.dtypes == bool])
new_categorical_cols.extend(["toss", "bat_first", "format"])

print(f"Currrent length of categorical columns : {len(new_categorical_cols)}")

Currrent length of categorical columns : 3


In [6]:
# Question B

# now build a decision tree to predict the result given the trainingData
# Step 0: 
from decisionTree import DecisionTreeModel
from utils import evaluate, plotA

maxDepths = [15, 25, 35, 45]
trainAccuracies = []
validationAccuracies = []
testAccuracies = []

categorical_cols = new_categorical_cols
continuous_cols = ["Unnamed", "year", "fow", "score", "rpo"]

for maxDepth in maxDepths:
    print(f"---------------Testing for maxDepth : {maxDepth}-------------------")
    model = DecisionTreeModel(maxDepth, categorical_cols = categorical_cols, continuous_cols = continuous_cols)

    # TRAIN THE MODEL WITH TRAIN DATA
    print("Training on the train data")
    model.fit(trainDataX, trainDataY)

    # if maxDepth == 5:
    #     model.print_tree()
    
    # PREDICT THE MODEL ON TRAIN DATA
    print("Prediction on train data")
    predictY_train = model.predict(trainDataX)
    
    # EVALUATION
    train_accuracy = evaluate(predictY_train, trainDataY)
    
    # PREDICT THE MODEL ON VALIDATION DATA
    print("Prediction on validation data")
    predictY_validation = model.predict(validationDataX)
    
    # EVALUATION
    validation_accuracy = evaluate(predictY_validation, validationDataY)
    
    # PREDICT THE MODEL ON Test DATA
    print("Prediction on test data")
    predictY_test = model.predict(testDataX)
    
    # EVALUATION
    test_accuracy = evaluate(predictY_test, testDataY)

    trainAccuracies.append(train_accuracy)
    validationAccuracies.append(validation_accuracy)
    testAccuracies.append(test_accuracy)

plotA(maxDepths, trainAccuracies, validationAccuracies, testAccuracies)

---------------Testing for maxDepth : 15-------------------
Training on the train data
Prediction on train data
 Accuracy : 0.7165, Precision: 0.6594, Recall: 0.9036, F1-Score: 0.7624
Prediction on validation data
 Accuracy : 0.5828, Precision: 0.5587, Recall: 0.7788, F1-Score: 0.6506
Prediction on test data
 Accuracy : 0.5822, Precision: 0.5414, Recall: 0.7707, F1-Score: 0.6360
---------------Testing for maxDepth : 25-------------------
Training on the train data
Prediction on train data
 Accuracy : 0.8526, Precision: 0.8137, Recall: 0.9170, F1-Score: 0.8623
Prediction on validation data
 Accuracy : 0.6080, Precision: 0.5936, Recall: 0.6797, F1-Score: 0.6337
Prediction on test data
 Accuracy : 0.6039, Precision: 0.5712, Recall: 0.6572, F1-Score: 0.6112
---------------Testing for maxDepth : 35-------------------
Training on the train data
Prediction on train data
 Accuracy : 0.9283, Precision: 0.8987, Recall: 0.9665, F1-Score: 0.9314
Prediction on validation data
 Accuracy : 0.5977, Pr

KeyboardInterrupt: 

| Max Depth | Train Accuracy | Validation Accuracy | Test Accuracy |
| --------- | -------------- | ------------------- | ------------- |
| 15        | 0.788          | 0.614               | 0.608         |
| 25        | 0.936          | 0.616               | 0.620         |
| 35        | 0.982          | 0.584               | 0.626         |
| 45        | 0.991          | 0.579               | 0.625         |

Training accuracy increases steadily as the maximum depth grows (from 0.78 → 0.99).
→ This shows the model is fitting the training data better as depth increases — a clear sign of overfitting.

Validation and test accuracies remain almost constant (around 0.58–0.63).
→ Even though the tree becomes more complex, it doesn’t generalize better on unseen data.

One-Hot Encoding slightly improves test performance compared to part (a).

In part (a): Test accuracy peaked at ~0.592

In part (b): Test accuracy improved to ~0.626
 Encoding categorical variables helped the model represent discrete features more effectively.

Beyond depth = 35, there’s no gain in validation/test accuracy → optimal depth seems to be around 30–35.

After applying one-hot encoding, the decision tree achieved slightly better generalization, with test accuracy increasing from ~0.59 (without encoding) to ~0.62. This improvement arises because one-hot encoding allows the model to better distinguish between categorical levels. However, increasing tree depth beyond 35 leads to overfitting — the training accuracy reaches nearly 100%, while test accuracy stagnates — indicating that deeper trees memorize the training data rather than learning generalizable patterns.